app/frontend/wallet/helpers/generateDeriveXpubFn.js

```diff
@@ -0,0 +1,38 @@
+const indexIsHardened = require('./indexIsHardened')
+const {derivePublic} = require('cardano-crypto.js')
+
+function generateDeriveXpubFn(deriveXpubHardenedFn, derivationScheme) {
+  const derivedXpubs = {}
+
+  async function deriveXpub(absDerivationPath) {
+    const memoKey = JSON.stringify(absDerivationPath)
+
+    if (!derivedXpubs[memoKey]) {
+      const deriveHardened =
+        absDerivationPath.length === 0 ||
+        indexIsHardened(absDerivationPath[absDerivationPath.length - 1])
+
+      /*
+      * TODO - reset cache if the promise fails, for now it does not matter
+      * since a failure (e.g. rejection by user) there leads to
+      * the creation of a fresh wallet and crypto provider instance
+      */
+      derivedXpubs[memoKey] = deriveHardened
```

Comment:
Why do we actually put *promise* into the map? If we do this because of concurrent requests, please add a comment there

Reply:
yes, otherwise you would end up firing multiple requests at Trezor when deriving multiple addresses.